### PR TITLE
#[UIC-1600] Fixed publish & subscriber sync issue

### PR DESCRIPTION
**Pull request by ankursinghal2005**
### CATEGORY

Choose one

- [X] Bug Fix
- [X] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Slice should open as popup only for related columns which Subscriber layer, "used as modal" selected.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@arpit-agarwal

### CODE REVIEW

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js [hunk 1]
```diff
@@ -2,15 +2,61 @@ import * as _ from 'lodash';
...
+export function getActionsMapForSlice(slice) {
+    let actions = {};
+    if (slice && slice.formData && slice.formData.hasOwnProperty('subscriber_layers')) {
+        let subscriberLayers = slice.formData.subscriber_layers
+        if (subscriberLayers) {
+            subscriberLayers.forEach(element => {
+                if (isUseAsModalActionExist(element.actions)) {
+                    if (!actions['USE_AS_MODAL'])
+                        actions['USE_AS_MODAL'] = [];
+                    actions['USE_AS_MODAL'] = _.union(actions['USE_AS_MODAL'], [element.linked_slice[0]['publisher_id']]);
```
Comment (arpit-agarwal) [inline superset/assets/src/dashboard/util/publishSubscriberUtil.js]: how we are ensuring that linked_slice will have only single element?

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js [hunk 2]
```diff
@@ -2,15 +2,61 @@ import * as _ from 'lodash';
...
+    }
+    return false;
+}
+
+export function getActionsMapForSlice(slice) {
+    let actions = {};
+    if (slice && slice.formData && slice.formData.hasOwnProperty('subscriber_layers')) {
+        let subscriberLayers = slice.formData.subscriber_layers
+        if (subscriberLayers) {
+            subscriberLayers.forEach(element => {
```
Comment (arpit-agarwal) [inline superset/assets/src/dashboard/util/publishSubscriberUtil.js]: rename `element` to `susbcriberLayer`

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js (hunk 1 above)
Comment (arpit-agarwal) [inline superset/assets/src/dashboard/util/publishSubscriberUtil.js]: can we simplify this to 
```
actions['USE_AS_MODAL'].PUSH([element.linked_slice[0]['publisher_id']]));
```

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js [hunk 3]
```diff
@@ -2,15 +2,61 @@ import * as _ from 'lodash';
...
+                if (isUseAsModalActionExist(element.actions)) {
+                    if (!actions['USE_AS_MODAL'])
+                        actions['USE_AS_MODAL'] = [];
+                    actions['USE_AS_MODAL'] = _.union(actions['USE_AS_MODAL'], [element.linked_slice[0]['publisher_id']]);
+                }
+                let subscribe_columns = element.linked_slice[0]['subscribe_columns']
+                subscribe_columns.forEach(item => {
```
Comment (arpit-agarwal) [inline superset/assets/src/dashboard/util/publishSubscriberUtil.js]: rename `item` to `column`

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js [hunk 4]
```diff
@@ -29,11 +75,14 @@ export function getUniqueActionsForSlice(slice) {
...
-        const subscribers = Object.values(publishSubscriberMap.subscribers);
-        var item = _.find(subscribers, function (item) {
-            return ((item.actions.indexOf(USE_AS_MODAL) > -1) && (Object.keys(item.linked_slices).indexOf(publisherId) != -1));
-        });
-        return item ? item.id : undefined;
+        const subscribers = publishSubscriberMap.publishers[publisherId]['subcribers'];
+        let modalSliceId;
+        subscribers.forEach(subscriberId => {
+            let subscriberSlice = publishSubscriberMap.subscribers[subscriberId];
+            if ((subscriberSlice.actions[USE_AS_MODAL] && (subscriberSlice.actions[USE_AS_MODAL].indexOf(parseInt(publisherId)) != -1)))
```
Comment (arpit-agarwal) [inline superset/assets/src/dashboard/util/publishSubscriberUtil.js]: we can use some function here isntead of indexOF

File: superset/assets/src/dashboard/components/Dashboard.jsx [hunk 5]
```diff
@@ -310,14 +310,11 @@ class Dashboard extends React.PureComponent {
...
     let keyExists = false;
-    if (publishers && publishers[filterKey]) {
-      const linked_slices = publishers[filterKey].subcribers;
-      const key = parseInt(sliceId);
-      if (linked_slices instanceof Array) {
-        keyExists = linked_slices.indexOf(sliceId) != -1
-      }
+    if (subscribers && subscribers[sliceId]) {
+      let applyFilterPublisherIds = subscribers[parseInt(sliceId)].actions['APPLY_FILTER'];
+      keyExists = (applyFilterPublisherIds && applyFilterPublisherIds.indexOf(parseInt(publisherId)) != -1) ? true : false;
```
[the remark this answers is not in the document]
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/components/Dashboard.jsx]: done

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js (hunk 2 above)
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/publishSubscriberUtil.js]: done

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js (hunk 1 above)
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/publishSubscriberUtil.js]: done

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js (hunk 3 above)
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/publishSubscriberUtil.js]: done

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js (hunk 4 above)
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/publishSubscriberUtil.js]: done

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js (hunk 1 above)
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/publishSubscriberUtil.js]: every subscriberLayer can subscribed only one linked slice

File: superset/assets/src/dashboard/util/publishSubscriberUtil.js [hunk 6]
```diff
@@ -2,15 +2,61 @@ import * as _ from 'lodash';
...
+    if (actions) {
+        if (actions.constructor == Object) {
+            return (actions[USE_AS_MODAL]) ? true : false;
+        }
+        else if (actions.constructor == Array) {
+            return (actions.indexOf(USE_AS_MODAL) != -1) ? true : false;
+        }
+    }
+    return false;
+}
```
[the remark this answers is not in the document]
Comment (ankursinghal2005) [reply superset/assets/src/dashboard/util/publishSubscriberUtil.js]: done

**Comment by arpit-agarwal** [comment]
LGTM

**Pull request merged**